### PR TITLE
Memory management changes

### LINKED ISSUE #3457: Arrow OutOfMemoryException in Local Auctionmark
Similar to our other Heap Space OOM (#3451), witnessed a `org.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.` error thrown while running auctionmark locally (with `cloud-benchmark:local`). This was actually caused by a lack of Direct Memory available, rather than Heap Memory.

## Details

Thrown running auctionmark locally on docker - within a volume attached.

Memory settings on Java task:
```
   - `"-Xmx1500m", "-Xms1500m", \
    "-XX:MaxDirectMemorySize=2000m", \
    "-XX:MaxMetaspaceSize=500m", \`
```

Node config:
```yaml
txLog: !Local
  path: /var/lib/xtdb/local-log

storage: !Local
  path: /var/lib/xtdb/local-buffers

metrics:
```

## Stack Trace

```
10:32:03.039 [xtdb.compactor-94] ERROR xtdb.compactor - Error running compaction job.
org.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.
	at io.netty.buffer.PooledByteBufAllocatorL.allocate(PooledByteBufAllocatorL.java:66)
	at org.apache.arrow.memory.netty.NettyAllocationManager.<init>(NettyAllocationManager.java:82)
	at org.apache.arrow.memory.netty.NettyAllocationManager.<init>(NettyAllocationManager.java:89)
	at org.apache.arrow.memory.netty.NettyAllocationManager$1.create(NettyAllocationManager.java:39)
	at org.apache.arrow.memory.BaseAllocator.newAllocationManager(BaseAllocator.java:367)
	at org.apache.arrow.memory.BaseAllocator.newAllocationManager(BaseAllocator.java:362)
	at org.apache.arrow.memory.BaseAllocator.bufferWithoutReservation(BaseAllocator.java:350)
	at org.apache.arrow.memory.BaseAllocator.buffer(BaseAllocator.java:328)
	at org.apache.arrow.memory.BaseAllocator.buffer(BaseAllocator.java:291)
	at org.apache.arrow.vector.BaseVariableWidthVector.reallocDataBuffer(BaseVariableWidthVector.java:576)
	at org.apache.arrow.vector.BaseVariableWidthVector.handleSafe(BaseVariableWidthVector.java:1344)
	at org.apache.arrow.vector.BaseVariableWidthVector.copyFromSafe(BaseVariableWidthVector.java:1446)
	at xtdb.vector.ScalarVectorWriter.rowCopier$lambda$1(FieldVectorWriters.kt:62)
	at xtdb.vector.StructVectorWriter.rowCopier$lambda$14(StructVectorWriter.kt:179)
	at xtdb.vector.DenseUnionVectorWriter$ChildWriter.rowCopier$lambda$0(DenseUnionVectorWriter.kt:108)
	at xtdb.vector.DenseUnionVectorWriter.duvRowCopier$lambda$6(DenseUnionVectorWriter.kt:209)
	at xtdb.compactor$__GT_reader__GT_copier$reader__GT_copier$reify__15783.copyRow(compactor.clj:54)
	at xtdb.compactor$merge_segments_into.invokeStatic(compactor.clj:95)
	at xtdb.compactor$merge_segments_into.invoke(compactor.clj:59)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15834$fn__15835$fn__15844$fn__15845.invoke(compactor.clj:137)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15834$fn__15835$fn__15844.invoke(compactor.clj:135)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15834$fn__15835.invoke(compactor.clj:135)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15834.invoke(compactor.clj:124)
	at xtdb.compactor$exec_compaction_job_BANG_.invokeStatic(compactor.clj:124)
	at xtdb.compactor$exec_compaction_job_BANG_.invoke(compactor.clj:119)
	at xtdb.compactor$eval15949$fn__15951$compactor_task__15980$fn__15982.invoke(compactor.clj:272)
	at xtdb.compactor$eval15949$fn__15951$compactor_task__15980.invoke(compactor.clj:270)
	at xtdb.compactor$eval15949$fn__15951$fn__15985$fn__15986$fn__15987$fn__15993.invoke(compactor.clj:307)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:309)
Caused by: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 268435456 byte(s) of direct memory (used: 1933574144, max: 2097152000)
	at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:843)
	at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:772)
	at io.netty.buffer.UnpooledUnsafeNoCleanerDirectByteBuf.allocateDirect(UnpooledUnsafeNoCleanerDirectByteBuf.java:30)
	at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeNoCleanerDirectByteBuf.allocateDirect(UnpooledByteBufAllocator.java:186)
	at io.netty.buffer.UnpooledDirectByteBuf.<init>(UnpooledDirectByteBuf.java:64)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.<init>(UnpooledUnsafeDirectByteBuf.java:41)
	at io.netty.buffer.UnpooledUnsafeNoCleanerDirectByteBuf.<init>(UnpooledUnsafeNoCleanerDirectByteBuf.java:25)
	at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeNoCleanerDirectByteBuf.<init>(UnpooledByteBufAllocator.java:181)
	at io.netty.buffer.UnpooledByteBufAllocator.newDirectBuffer(UnpooledByteBufAllocator.java:91)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
	at io.netty.buffer.PooledByteBufAllocatorL$InnerAllocator.newDirectBufferL(PooledByteBufAllocatorL.java:167)
	at io.netty.buffer.PooledByteBufAllocatorL$InnerAllocator.directBuffer(PooledByteBufAllocatorL.java:210)
	at io.netty.buffer.PooledByteBufAllocatorL.allocate(PooledByteBufAllocatorL.java:57)
```

## TODO

- [ ] Investigate sources/places that we allocate direct memory.
- [ ] Run locally with a profiler and see behaviour.

### LINKED ISSUE #3464: Replacing ArrowBufLRU with Caffeine cache
There are some overflowing issues with the ArroBufLRU. Replace the ArrowBufLRU cache with a memory bound Caffeine cache.

### PR DESCRIPTION
**Pull request by FiV0**
- Brings in Caffeine instead of using our own ArrowBufLRU. ~The problem with Caffeine is that pinning and the actual logic (does an entry just go through the cache or is it actually cached) is not very transparent, hence the need to add extra logic around caffeine for `retain` handeling.~ This is fixed with the thread-safe `asMap` + `compute` approach now.
- Limit the number of compactor threads based on number of processors and direct memory available. We estimate conservatively that an compactor thread might need up to 800mb and that the indexer needs at least 400mb. So the number of compactor threads is at least 1 and at most half the number of processors available and at most one per 800mb of available direct memory (accounting for the indexer). 

Todo:
- [x] remove anything `ArrowBufLRU` related

### CODE REVIEW

File: core/src/main/clojure/xtdb/buffer_pool.clj [hunk 1]
```diff
@@ -36,25 +36,24 @@
...
     (util/->path (str "v" (util/->lex-hex-string version)))))
 
-(defn- free-memory [^Map memory-store]
-  (locking memory-store
-    (run! util/close (.values memory-store))
-    (.clear memory-store)))
+(defn- free-memory [^Cache memory-store]
+  ;; closing should happen via the remove method
+  (.invalidateAll memory-store)
+  (Thread/sleep 100))
```
Comment (FiV0) [inline core/src/main/clojure/xtdb/buffer_pool.clj]: I spend more time than I am willing to admit trying to get rid of this sleep. Even though we are using a synchronous cache (I think) the invalidation somehow happens async (it seems). Getting at least allocator problems when I don't add this.

Comment (jarohen) [reply core/src/main/clojure/xtdb/buffer_pool.clj]: Might `Cache/.cleanUp` after the invalidation do this?

File: core/src/main/clojure/xtdb/buffer_pool.clj [hunk 2]
```diff
@@ -401,8 +415,9 @@
 
   EvictBufferTest
   (evict-cached-buffer! [_ k]
-    (doto (.remove memory-store k)
-      (util/close)))
+    (when-let [buf (.getIfPresent memory-store k)]
+      (.invalidate memory-store k)
+      (util/close buf)))
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/buffer_pool.clj]: might this be a double-close? I'm assuming the invalidation will cause a close too

File: core/src/main/clojure/xtdb/util.clj [hunk 3]
```diff
@@ -663,3 +664,14 @@
                           (cons x (step (rest s) (conj seen v)))))))
                   xs seen)))]
      (step coll #{}))))
+
+(defn max-direct-memory
+  "Returns the maximum direct memory supposed to be used by the system"
+  []
+  (if-let [size-string (get (System/getProperties) "MaxDirectMemory")]
+    (let [^long size (parse-long (subs size-string 0 (dec (count size-string))))]
+      (case (last size-string)
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/util.clj]: nit: may need to lower-case this - I think you can specify it with upper-case too

File: core/src/main/clojure/xtdb/buffer_pool.clj (hunk 1 above)
Comment (FiV0) [reply core/src/main/clojure/xtdb/buffer_pool.clj]: No, doesn't help.

File: core/src/main/clojure/xtdb/util.clj (hunk 3 above)
Comment (FiV0) [reply core/src/main/clojure/xtdb/util.clj]: Btw this is a property we are specifying, not `-XX:MaxDirectMemorySize`.

File: core/src/main/clojure/xtdb/buffer_pool.clj (hunk 2 above)
Comment (FiV0) [reply core/src/main/clojure/xtdb/buffer_pool.clj]: I think the reason I added it (although not consciously), is that you run into the same issue as with `invalidateAll` above (i.e. the `removalListener` running somehow asynchronously).

File: core/src/main/clojure/xtdb/buffer_pool.clj (hunk 1 above)
Comment (FiV0) [reply core/src/main/clojure/xtdb/buffer_pool.clj]: Waiting on the `ForkJoinPool/commonPool` now. This is where things get executed asynchronously.

File: core/src/main/clojure/xtdb/buffer_pool.clj (hunk 2 above)
Comment (FiV0) [reply core/src/main/clojure/xtdb/buffer_pool.clj]: See above.

File: core/src/main/clojure/xtdb/util.clj (hunk 3 above)
Comment (danmason) [reply core/src/main/clojure/xtdb/util.clj]: Just out of curiosity - Is this being used anywhere, currently?

Comment (FiV0) [reply core/src/main/clojure/xtdb/util.clj]: Yes, I use it on the AWS runs.

**Review by jarohen — APPROVED** [review]

File: core/src/main/clojure/xtdb/buffer_pool.clj (hunk 1 above)
Comment (jarohen) [reply core/src/main/clojure/xtdb/buffer_pool.clj]: Yep - as discussed, pragmatically let's run with this for now until we can investigate whether there's something we can hook into in Caffeine. Might be an idea to look into their tests to see what they do?

**Pull request merged**